Make email validator case-insensitive

nospame opened the pull request:
## Product Description
Addresses the issue of capital letters in an email address potentially causing a validation error when inviting a web user. Previously, any email address typed with the name portion ending in a capital letter or the domain portion containing any capital letters would be marked invalid, e.g.
![image](https://github.com/user-attachments/assets/ece53fa1-cb16-45c2-b651-1d2fcc1e4b81)

![image](https://github.com/user-attachments/assets/ccec7234-93b6-4942-a0ce-88efe8e3fdf3)

Both of which are now valid.

## Technical Summary
[SAAS-15628](https://dimagi.atlassian.net/browse/SAAS-15628)
Updates the regular expression for email in `validators.ko` to allow capital letters. Also lowercases the email address as part of the web user invitation form's `clean_email`, which while not required here because of the `AdminInvitesUserFormValidator`, makes it consistent with other `clean_email` methods.

## Safety Assurance

### Safety story
This is a small change to make email addresses validation more accurate and not block the entry of some valid email addresses. I think the risks are quite low.

### Automated test coverage
We don't have tests for this monster of a regular expression.

### QA Plan
Not planning it.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-15628]: https://dimagi.atlassian.net/browse/SAAS-15628?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Make email validator case-insensitive

nospame committed:
Lowercase new web user email before validating

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/bootstrap5/validators.ko.js:
```diff
@@ -16,7 +16,7 @@ hqDefine("hqwebapp/js/bootstrap5/validators.ko", [
         validator: function (val) {
             if (val === undefined || val.length === 0) {return true;}  // do separate validation for required
-            var re = /(?:[a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+\/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21- ...
+            var re = /(?:[A-Za-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\.[A-Za-z0-9!#$%&'*+\/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?\.)+[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[A-Za-z0-9-]*[A-Za-z0-9]:(?:[\ ...
```
Alternatively, you could run `test` on `value.toLowerCase()` and make the regex a smidge simpler. But it's already a monstrosity, so this is just a suggestion.

orangejenny reviewed (approved):
Tests *would* be pretty nice given the complexity of the regex.

If you extracted the regex into a constant somewhere, I think it'd be pretty straightforward to set up utility tests, [assert_properties.js](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqwebapp/static/hqwebapp/spec/assert_properties_spec.js) is a nice example. But I defer to you on whether or not you want to increase the scope of this, this change looks fine.

nospame commented:
> Tests _would_ be pretty nice given the complexity of the regex.

@orangejenny That's totally reasonable. Worked on it a bit this afternoon, and actually having a bit of a time deciphering what this regex does relative to the actual email standard (with the examples given on Wikipedia, at least). It seems like it's actually a fair bit less restrictive in some senses, and more strict in others. 

Would you think it makes more sense to write tests according to what the expression actually does and retain most of existing behavior, or write them according to what the standard "should be" and change the expression to fit?

orangejenny commented:
@nospame I'd write them to document the current behavior and not change the functionality. I'm curious what kinds of differences there are, but that's just idle curiosity talking.

nospame committed:
Lint email validation regex

nospame committed:
Move email validation regex to constants

nospame committed:
Add email_validator_spec

nospame commented:
@orangejenny In general it's just not as explicit as the actual standard, which is almost impossibly comprehensive.

Moved the expression to constants c39df56, Added an `email_validator_spec` 453f37a, also updated and force-pushed the lint commit e92b8b3 because I'd accidentally reverted the expression to its older version.

nospame committed:
"Bootstrap 5 Migration - Rebuilt diffs"

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/spec/email_validator_spec.js:
```diff
@@ -0,0 +1,83 @@
+        });
+
+        it('should allow specified special characters in the local part', function () {
+            assert.ok(testEmail('.\x01!#$%&"\'*+/=?^_`{|}~-@example.com'));
```
wild

nospame replied:
Agreed! I think this is one of the places where, strictly speaking, some of these characters should only be allowed based on context rather than all the time, e.g. a quote should always be paired. But the existing expression doesn't test for that.

orangejenny reviewed (approved):
this looks great
